src/encoder/mod.rs

```diff
@@ -0,0 +1,425 @@
+use std::io::{Write, Seek};
+use std::collections::BTreeMap;
+use byteorder::{NativeEndian};
+
+use crate::decoder::ifd::Tag;
+use crate::error::{TiffResult, TiffFormatError, TiffError};
+
+mod writer;
+pub mod colortype;
+
+use self::writer::*;
+use self::colortype::*;
+
+/// Type to represent tiff values of type `RATIONAL`
+pub struct Rational {
+    pub n: u32,
+    pub d: u32,
+}
+
+/// Trait for types that can be encoded in a tiff file
+pub trait TiffValue {
+    const BYTE_LEN: u32;
+    const FIELD_TYPE: u16;
```

Comment:
The decoder already has a type for this:  `decoder::ifd::Type`. It might be clearer to use that than hardcode `FIELD_TYPE` values in, it only lacks a way to retrieve the variant values but you can easily add a `ToPrimitive` derive since we already depend on `num-derive`.

Reply:
You mean like this right?
```rust
const FIELD_TYPE: ifd::Type;
```
or do you mean removing `FIELD_TYPE` and add a `ifd::Type` parameters to the other functions?

Reply:
Yep i mean keeping `FIELD_TYPE` but with type `ifd::Type`. I think it makes the code slightly clearer to read (e.g. `Type::BYTE` instead of `1`) and it also makes sure that `ifd::Type` is the only place where we can possibly get the magic numbers wrong.